util/tls/tls_socket.cc

```diff
@@ -206,14 +206,21 @@ io::Result<size_t> TlsSocket::WriteSome(const iovec* ptr, uint32_t len) {
   DCHECK(engine_);
   DCHECK_GT(len, 0u);
 
-  Engine::Buffer dest{reinterpret_cast<uint8_t*>(ptr->iov_base), ptr->iov_len};
+  Engine::Buffer dest;
   size_t send_total = 0;
 
+  if (len == 1) {
+    dest = {reinterpret_cast<uint8_t*>(ptr->iov_base), ptr->iov_len};
+  } else {
+    write_buffer_.Clear();
+    for (size_t i = 0; i < len; i++) {
```

Comment:
what if we got 1MB chunks? 
I think this logic makes sense only when total length of the iovec is small. otherwise you copy your data twice - it's too costly.

Reply:
looking at the code further, I actually think better solution would be is to avoid calling `MaybeSendOutput` in some cases and buffer already encrypted data inside tls engine.  So right now we call `MaybeSendOutput` after each call to `engine_->Write(dest);`  (i.e. prematurely) but instead we could call when tls enginer buffers are at significant capacity (or full).

Reply:
for that you need to reorder the logic that handles op_result and calls MaybeSendOutput because you must know if there is more data you potentially need to write.

Reply:
Fair point about the total length of the iovec, but we should still buffer calls to `Write` so we don't create too many TLS records

Reply:
Are you saying each call to tls engine creates a separate record?

Reply:
Pretty sure, yeah (the overhead is around 20-30 bytes IIUC but I haven't tested it)

Reply:
I found this related thread: https://mta.openssl.org/pipermail/openssl-users/2020-February/011753.html
in that case, a reasonable behaviour would be to continue copying until you reach a size larger than `max tls frame size * alpha` and then flush and continue. A quick google search gave 16K as an answer so I think 4kb or 8kb is a good threshold.  In addition, I do not think iobuf is needed here - you can know in advance how much memory you need (4kb or less) and reserve accordingly. a simple `vector<uint8_t>` should suffice I think.

Reply:
I ended up doing a smaller size - 1408 - because of the usual MTU size of 1500 plus space for metadata. Also wanted to reduce memory footprint of a connection so I just moved the buffer to the stack. LMK if you think it's unreasonable.